### PR TITLE
odin 2022-05

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Odin changed their pre-1.0 version numbering to `dev-YYYY-MM`. This is supposed to be new versioning strategy for 0.13.1+ releases until official 1.0 where upstream will switch to semantic versions (may require a `version_scheme` change then).

References:
- https://github.com/odin-lang/Odin/discussions/916
- https://github.com/odin-lang/Odin/issues/887

### CODE REVIEW

**Comment by cho-m** [comment]
Needs `llvm@11` on Linux, `llvm@13` on M1 mac, and one of `llvm@{11,12,13}` on Intel mac.

File: Formula/odin.rb
Comment (cho-m) [inline Formula/odin.rb:18]: Using `llvm` will probably break with `llvm@14` #97618.
So, PR will be blocked until `llvm` PR is merged and then `llvm@13` is added.

**Comment by Bo98** [comment]
> Needs `llvm@11` on Linux, `llvm@13` on M1 mac, and one of `llvm@{11,12,13}` on Intel mac.

Why? Seems weird to have varying dependency trees.

**Comment by cho-m** [comment]
> > Needs `llvm@11` on Linux, `llvm@13` on M1 mac, and one of `llvm@{11,12,13}` on Intel mac.
> 
> Why? Seems weird to have varying dependency trees.

That's how the upstream has set this up.
From the Makefile https://github.com/odin-lang/Odin/blob/dev-2022-03/Makefile

macOS:
```make
        ifeq ($(ARCH), arm64)
            $(error "Requirement: llvm-config must be base version 13 for arm64")
        else 
            $(error "Requirement: llvm-config must be base version greater than 11 for amd64/x86")
        endif 
```

Linux:
```make
        ifeq ($(shell $(LLVM_CONFIG) --version | grep '^11\.'),)
            $(error "Requirement: llvm-config must be version 11")
        endif
```

**Comment by Bo98** [comment]
Seems odd since `build_odin.sh` (after the tag) does not have that restriction.

**Comment by cho-m** [comment]
The master branch still has the restrictions, e.g. https://github.com/odin-lang/Odin/blob/master/build_odin.sh
```sh
	# allow for arm only llvm's with version 13
	if [ ARCH == arm64 ]; then
		MIN_LLVM_VERSION=("13.0.0")
	else
		# allow for x86 / amd64 all llvm versions begining from 11
		MIN_LLVM_VERSION=("11.1.0")
	fi
```

It looks like they are seeing issues with particular LLVM releases and thus restricting the ones they allow. Some related refs:
- https://github.com/odin-lang/Odin/pull/1380
- https://github.com/odin-lang/Odin/discussions/1038

**Comment by Bo98** [comment]
min != exact

**Comment by cho-m** [comment]
Yeah, taking another look, it is only checking min bound. Still, the Makefile is restricted, so we cannot use another LLVM version unless it is modified.

Documentation still points to using `make` and main usage of `build_odin.sh` is within CI that installs explicit Ubuntu `llvm-11 clang-11` / Homebrew `llvm@11`.

Would need upstream comment on how coupled the LLVM version is with Odin. Though, given ARM builds are on `llvm@13` and Windows builds are on `llvm@12`, it seems odd that Linux is only left at `llvm@11`.

**Comment by cho-m** [comment]
Looks like upstream has now updated Makefile to use build script without any upper bound contraint on `llvm`.

A fairly large number of commits, so may be better to wait for `2022-04` which should be released "soon".

**Comment by cho-m** [comment]
Build script brings in its own complications since it uses current date for version, so any rebuild is only valid in same month of release.
```
-DODIN_VERSION_RAW=\"dev-$(date +"%Y-%m")\"
```

**Comment by Bo98** [comment]
The makefile had the same issue, but it sounds like a good bug report or pull request to make to allow that to be overridable.

**Comment by cho-m** [comment]
`__ulock_wait2` might not be available in 10.15 and older. Re-running with `CI-no-fail-fast` to check. May need to restrict to macOS 11+

https://github.com/odin-lang/Odin/blob/dev-2022-05/core/sync/futex_darwin.odin#L33
```odin
	s := __ulock_wait2(UL_COMPARE_AND_WAIT | ULF_NO_ERRNO, f, u64(expected), timeout_ns, 0)
```

File: Formula/odin.rb
Comment (chenrui333) [inline Formula/odin.rb:7]: so the version scheme got changed to CalVer?

File: Formula/odin.rb
Comment (chenrui333) [inline Formula/odin.rb:22]: 👍

**Review by chenrui333 — APPROVED** [review]

File: Formula/odin.rb
Comment (cho-m) [reply Formula/odin.rb:7]: At least during the beta (pre-1.0) phase. Upstream plans to switch to semver again once 1.0 is released (https://github.com/odin-lang/Odin/issues/887).

Comment (chenrui333) [reply Formula/odin.rb:7]: I have some concern that we will run into `version_scheme` issue down the path.

**Pull request closed without merging**
